tool/ood-gen/lib/planet.ml

```diff
@@ -51,65 +40,96 @@ let validate_entries entries =
       { entry with id; authors })
     entries
 
-let scrape () =
-  let feeds = feeds () in
+let fetch_feed (source : River.source) =
+  try Some (source, River.fetch source)
+  with e ->
+    print_endline
+      (Printf.sprintf "failed to scrape %s: %s" source.name
+         (Printexc.to_string e));
+    None
+
+let scrape_feed ((source : River.source), (feed : River.feed)) =
+  let posts = River.posts [ feed ] in
+  let name = River.name feed in
+  let entries = posts |> River.create_atom_entries |> validate_entries in
+  let updated = Ptime.of_float_s (Unix.gettimeofday ()) |> Option.get in
+  let atom_feed =
+    Syndic.Atom.feed ~id:(Uri.of_string source.url)
+      ~title:(Text (River.name feed))
+      ~updated entries
+  in
+  Syndic.Atom.write atom_feed ("data/planet/" ^ name ^ ".xml");
+  posts
+  |> List.iter (fun (post : River.post) ->
+         let title = River.title post in
+         let slug = Utils.slugify title in
+         let output_file = "data/planet/" ^ name ^ "/" ^ slug ^ ".md" in
+         if Sys.file_exists output_file then
+           print_endline
+             (Printf.sprintf "%s/%s already exist, not scraping again" name slug)
+         else
+           let oc = open_out output_file in
+           let content = River.content post in
+           let url = River.link post in
+           let date = River.date post |> Option.map Syndic.Date.to_rfc3339 in
+           match (url, date) with
+           | None, _ ->
+               print_endline
+                 (Printf.sprintf "skipping %s/%s: item does not have a url" name
+                    slug)
+           | _, None ->
+               print_endline
+                 (Printf.sprintf "skipping %s/%s: item does not have a date"
+                    name slug)
+           | Some url, Some date ->
+               let url = Uri.to_string url in
+               let preview_image = River.seo_image post in
+               let description = River.meta_description post in
+               let author = River.author post in
+               let metadata =
+                 {
+                   title;
+                   url = Some url;
+                   date;
+                   preview_image;
+                   description;
+                   featured = None;
+                   authors = Some [ author ];
+                 }
+               in
+               let s = Format.asprintf "%a\n%s\n" pp_meta metadata content in
+               Printf.fprintf oc "%s" s;
```

Comment:
```suggestion
         let ( let* ) (opt, err) some =
           Option.fold ~none:(Printf.printf "%s/%s: %s" name slug err) ~some opt
         in
         let* _ =
           ( (if Sys.file_exists output_file then None else Some ()),
             "already exist, not scraping again" )
         in
         let oc = open_out output_file in
         let content = River.content post in
         let* url = (River.link post, "skipping, item does not have a url") in
         let date = River.date post |> Option.map Syndic.Date.to_rfc3339 in
         let* date = (date, "skipping, item does not have a date") in
         let url = Some (Uri.to_string url) in
         let preview_image = River.seo_image post in
         let description = River.meta_description post in
         let authors = Some [ River.author post ] in
         let metadata =
           {
             title;
             url;
             date;
             preview_image;
             description;
             featured = None;
             authors;
           }
         in
         let s = Format.asprintf "%a\n%s\n" pp_meta metadata content in
         Printf.fprintf oc "%s" s;
         close_out oc)
```

Reply:
Hm, one thing I greatly value is not to use the same operator name for different things. We're already using `let*` from `lwt` so that name is taken. As I see it, it is quite sad that we can't give custom let binding operators proper names - the set of characters we can use is so limited. :thinking:

I think with these kinds of operators, we have to carefully weigh whether the reduced obviousness of the code is worth it. The short-circuiting 404-binding-operator in the handlers makes sense to me: that one is getting enough use to make teaching the concept worthwhile, and it eliminates quite a bit of nesting for some handlers (which would otherwise have to be split in different functions to stay readable).

I agree that the nesting is bad. In this case, I'm moving the code into a separate function `scrape_post` to maintain a nesting of 3 levels.